### PR TITLE
Separated event from snapshot storage and started using the value `snapshottedEventCreatedAt` in snapshots to initialize the database sort key.

### PR DESCRIPTION
**Pull request by javiertoledo**
## Description

TL;DR. This change makes it possible to build the correct state of an entity starting from an arbitrary snapshot in the database, making the process of snapshot generation idempotent, and allowing snapshot writes from competing threads without inducing risks of missing events.

In the current codebase, events and entity snapshots are ordered in the database by a timestamp calculated at the time of persisting them. This makes Booster apps especially sensible to insertion order, making it harder to properly cache snapshots at read time, and forcing us to limit snapshot creation to the events dispatcher. While this approach has shown great resiliency, it comes at the cost of performance. When a certain number of commands and event handlers need to access the most recent version of an entity to validate further actions, and there's no recent snapshot, all of them have to recalculate the last state from the event source, making the event dispatcher process slower.

This behavior has been changed in #1116, where @gonzalogarciajaubert pointed out about a potential issue caused by the way we were setting up the snapshots sort keys. This PR fixes that potential issue with a rework on how events and snapshots are stored and ordered in the database.

> Notice that this change has been worked on top of the changes in #1116, so they're related and must be merged in order.

## Checks
- [x] Project Builds
- [x] Project passes tests and checks
~- [ ] Updated documentation accordingly~

## Additional checklist (Scenarios to add to the automated tests)

- [x] An exception while reducing events from a specific entity does no longer affect the reduction of other entities (#1254)
- [x] Event handlers are processed in parallel with read model updates
- [x] An error reducing an entity halts any future reduction for the failing entity (until someone fixes it)
- [x] An error thrown in an event handler does not stop other event handlers
- [x] When a snapshot already exists in the database, trying to rewrite it fails silently (writing it again is not needed)
- [x] Once a failing entity is fixed, the process restarts where it was left no matter how much time passed
- [x] Event handlers no longer wait for snapshots and read models to be calculated, they are processed in parallel, and a failure in one of the threads does not affect the other
- [ ] This change increases performance enough to augment the current load test limits

### CODE REVIEW

File: packages/framework-types/src/provider.ts [hunk 1]
```diff
@@ -35,16 +37,22 @@ export interface ProviderEventsLibrary {
...
     limit: number,
     afterCursor: Record<string, string> | undefined,
     entityTypeName: string
   ): Promise<PaginatedEntitiesIdsResult>
-  /** Streams an event to the corresponding event handler */
-  store(eventEnvelopes: Array<EventEnvelope>, config: BoosterConfig): Promise<void>
+  /** Stores a list of events in the events store */
+  store(eventEnvelopes: Array<NonPersistedEventEnvelope>, config: BoosterConfig): Promise<void>
+  /** Stores an entity snapshot in the event store */
+  storeSnapshot(snapshotEnvelope: EntitySnapshotEnvelope, config: BoosterConfig): Promise<void>
```
Comment (javiertoledo) [inline packages/framework-types/src/provider.ts:55]: This PR introduces a dedicated method in the `ProviderLibrary` interface to store snapshots.

File: packages/framework-types/src/envelope.ts [hunk 2]
```diff
@@ -26,18 +26,28 @@ export interface ScheduledCommandEnvelope extends Envelope {
...
+}
+export interface EventEnvelope extends NonPersistedEventEnvelope {
+  persistedAt: string
+}
+
+export interface EntitySnapshotEnvelope extends EventStoreEntryEnvelope {
+  kind: 'snapshot'
+  snapshottedEventCreatedAt: string
+  snapshottedEventPersistedAt: string
+}
```
Comment (javiertoledo) [inline packages/framework-types/src/envelope.ts:50]: There's a new type semantics for event store entries:
* `NonPersistedEventEnvelope`: An event that has not been stored yet
* `EventEnvelope`: An event that has been stored (it has a `persistedAt` timestamp set)
* `EntitySnapshotEnvelope`: A dedicated snapshot envelope that forces the presence of `snapshottedEventCreatedAt` and `snapshottedEventPersistedAt`. These fields reference the latest event processed for the current snapshot.

File: packages/framework-provider-aws/src/library/events-adapter.ts [hunk 3]
```diff
@@ -101,28 +103,78 @@ export async function storeEvents(
...
+        TableName: config.resourceNames.eventsStore,
+        ConditionExpression: `${eventsStoreAttributes.partitionKey} <> :partitionKey AND ${eventsStoreAttributes.sortKey} <> :sortKey`,
+        ExpressionAttributeValues: {
+          ':partitionKey': partitionKey,
+          ':sortKey': sortKey,
+        },
+        Item: {
+          ...snapshotEnvelope,
+          [eventsStoreAttributes.partitionKey]: partitionKey,
+          [eventsStoreAttributes.sortKey]: sortKey,
```
Comment (javiertoledo) [inline packages/framework-provider-aws/src/library/events-adapter.ts:136]: One detail here: While we're introducing a new `snapshottedEventPersistedAt` field and using it as the sort key value, the `eventsStoreAttributes.sortKey` constant is still `createdAt`. This means that at the time of writing, the value stored will override whatever value we set in `createdAt` and replace It with the value set in `snapshottedEventPersistedAt`. I think that this is important to keep backward compatibility, but I'm totally open to other thoughts on this.

File: packages/framework-provider-aws/src/library/events-adapter.ts [hunk 4]
```diff
@@ -101,28 +103,78 @@ export async function storeEvents(
...
         ExpressionAttributeValues: {
           ':partitionKey': partitionKey,
-          ':sortKey': sortKey,
+          ':sortKey': persistableEvent.persistedAt,
         },
         Item: {
           ...eventEnvelope,
           [eventsStoreAttributes.partitionKey]: partitionKey,
-          [eventsStoreAttributes.sortKey]: sortKey,
+          [eventsStoreAttributes.sortKey]: persistableEvent.persistedAt,
```
Comment (javiertoledo) [inline packages/framework-provider-aws/src/library/events-adapter.ts:177]: Same here: Doing this we're overriding the `createdAt` that comes in the `eventEnvelope`by the calculated `persistedAt` value (see line 164 in this same file)

File: packages/framework-provider-azure/src/library/events-adapter.ts [hunk 5]
```diff
@@ -72,35 +78,69 @@ export async function readEntityLatestSnapshot(
...
-        ...eventEnvelope,
+        ...persistableEvent,
         [eventsStoreAttributes.partitionKey]: partitionKeyForEvent(
           eventEnvelope.entityTypeName,
-          eventEnvelope.entityID,
-          eventEnvelope.kind
+          eventEnvelope.entityID
         ),
-        [eventsStoreAttributes.sortKey]: new Date().toISOString(),
+        [eventsStoreAttributes.sortKey]: persistableEvent.persistedAt,
```
Comment (javiertoledo) [inline packages/framework-provider-azure/src/library/events-adapter.ts:111]: Same as in AWS, we're keeping the same sort key in the database to ease upgrades and backward compatibility

File: packages/framework-core/src/booster-event-dispatcher.ts [hunk 6]
```diff
@@ -38,9 +38,16 @@ export class BoosterEventDispatcher {
...
+      await Promise.allSettled([
+        await BoosterEventDispatcher.snapshotAndUpdateReadModels(
+          config,
+          entityName,
+          entityID,
+          eventStore,
+          readModelStore
+        ),
+        await BoosterEventDispatcher.dispatchEntityEventsToEventHandlers(eventEnvelopes, config),
+      ])
```
Comment (javiertoledo) [inline packages/framework-core/src/booster-event-dispatcher.ts:50]: With the changes introduced in this PR, and given that at this point all events are 100% persisted and settled in the database, it does no longer matter the order in which we calculate the read models and process event handlers, so we can now process them in parallel.

File: packages/framework-provider-aws/src/library/events-adapter.ts [hunk 7]
```diff
@@ -101,28 +103,78 @@ export async function storeEvents(
...
-  eventEnvelope: EventEnvelope
+  eventEnvelope: NonPersistedEventEnvelope
 ): Promise<void> {
   try {
-    const partitionKey = partitionKeyForEvent(eventEnvelope.entityTypeName, eventEnvelope.entityID, eventEnvelope.kind)
+    const partitionKey = partitionKeyForEvent(eventEnvelope.entityTypeName, eventEnvelope.entityID)
     // Generate a new timestamp as sorting key on every try until we can persist the event.
     // This way we guarantee ordering even with events that are stored in the same millisecond
-    const sortKey = new Date().toISOString()
+    const persistableEvent: EventEnvelope = {
```
Comment (gonzalogarciajaubert) [inline packages/framework-provider-aws/src/library/events-adapter.ts:162]: I think this object is never used

File: packages/framework-provider-aws/src/library/events-adapter.ts (hunk 3 above)
Comment (gonzalogarciajaubert) [reply packages/framework-provider-aws/src/library/events-adapter.ts:136]: Not sure if we will need it or not, but we are not going to persist a timestamp with the snapshot creation, are we?

File: packages/framework-provider-azure/src/library/events-adapter.ts [hunk 8]
```diff
@@ -72,35 +78,69 @@ export async function readEntityLatestSnapshot(
...
+  await cosmosDb
+    .database(config.resourceNames.applicationStack)
+    .container(config.resourceNames.eventsStore)
+    .items.create({
+      ...snapshotEnvelope,
+      [eventsStoreAttributes.partitionKey]: partitionKey,
+      [eventsStoreAttributes.sortKey]: sortKey,
+    })
+
+  logger.debug('[EventsAdapter#storeSnapshot] Snapshot stored')
```
Comment (gonzalogarciajaubert) [inline packages/framework-provider-azure/src/library/events-adapter.ts:145]: Could we improve this log?

```suggestion
  logger.debug('Snapshot stored', snapshotEnvelope)
```

File: packages/framework-provider-aws/src/library/events-adapter.ts (hunk 3 above)
Comment (javiertoledo) [reply packages/framework-provider-aws/src/library/events-adapter.ts:144]: I rolled back the introduction of the `persistedAt` field to use the `createdAt` field again, but I'm still updating the `createdAt` to match the one from the entity. As we're now generating snapshots from multiple threads, this is the only way to know if a specific snapshot has been generated before, but it could make sense to also keep an extra timestamp with the snapshot persist time. I'll add that field here.

**Comment by javiertoledo** [comment]
Merging this branch into #1116 to start working on **the merge™**
